x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -0,0 +1,216 @@
+                    @SuppressWarnings("unchecked")
+                    List<String> ids = (List<String>) a[1];
+                    return new PinnedQueryBuilder(organicQuery, ids.toArray(new String[0]));
+                }
+             );
+    static {
+        PARSER.declareObject(constructorArg(), (p, c) -> parseInnerQueryBuilder(p), ORGANIC_QUERY_FIELD);        
+        PARSER.declareStringArray(constructorArg(), IDS_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    public static PinnedQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            return PARSER.apply(parser, null);
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
+        if (ids.size() == 0 && organicQuery == null) {
+            return new MatchNoneQueryBuilder();
+        }
+        return this;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        MappedFieldType idField = context.fieldMapper(IdFieldMapper.NAME);
+        if (idField == null) {
+            return new MatchNoDocsQuery("No mappings");
+        }
+        if (this.ids.isEmpty()) {
+            return new CappedScoreQuery(organicQuery.toQuery(context), MAX_ORGANIC_SCORE);
+        } else {
+            BooleanQuery.Builder pinnedQueries = new BooleanQuery.Builder();
+
+            // Ensure each pin order using a Boost query with the relevant boost factor
+            int minPin = NumericUtils.floatToSortableInt(MAX_ORGANIC_SCORE) + 1;
+            int boostNum = minPin + ids.size();
+            float lastScore = Float.MAX_VALUE;
+            for (String id : ids) {
+                float pinScore = NumericUtils.sortableIntToFloat(boostNum);
+                assert pinScore < lastScore;
+                lastScore = pinScore;
+                boostNum--;
+                // Ensure the pin order using a Boost query with the relevant boost factor
+                Query idQuery = new BoostQuery(new ConstantScoreQuery(idField.termQuery(id, context)), pinScore);
+                pinnedQueries.add(idQuery, BooleanClause.Occur.SHOULD);
+            }
+
+            // Score for any pinned query clause should be used, regardless of any organic clause score, to preserve pin order.
+            // Use dismax to always take the larger (ie pinned) of the organic vs pinned scores
+            List<Query> organicAndPinned = new ArrayList<>();
+            organicAndPinned.add(pinnedQueries.build());
+            // Cap the scores of the organic query
+            organicAndPinned.add(new CappedScoreQuery(organicQuery.toQuery(context), MAX_ORGANIC_SCORE));
+            return new DisjunctionMaxQuery(organicAndPinned, 0);
```

Comment:
FYI seeing this made me notice that disjunction max queries are a bit inefficient at retrieving top hits at the moment, even though there are some easy wins, especially when the tie-break multiplier is 0, so I opened https://issues.apache.org/jira/browse/LUCENE-8922